### PR TITLE
fix: create equal sized batches within batch size for logging #2341

### LINKED ISSUE #2341: create equally sized batches before/during `rg.log`
**Is your feature request related to a problem? Please describe.**
I just wanted to upload 35K records and logging is not as robust as I would like.

**Describe the solution you'd like**
Ideally, I would like to divide records based on the allowed batch size, while equally dividing the text lengths over these batches. IMO, this could already alleviate some of the problems relatively easily. And this might also help with assessing and analyzing payload limitations(each payload is roughly equal in this case).

**Describe alternatives you've considered**
N.A.

**Additional context**
N.A.

### PR DESCRIPTION
**Pull request by davidberenstein1957**
# Description

dives logging data in equal batches in two ways
1. len(record) ==700 AND batch_size == 500 => batch_size = 350
2. if size of record [2,2,1,1] batches 0=>[[2,1][2,1]]

Closes #2341 
**Type of change**

(Please delete options that are not relevant. Remember to title the PR according to the type of change)

- [X] New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And ideally, reference `tests`)

Not sure how to test this

**Checklist**
N.A.

### CODE REVIEW

**Comment by tomaarsen** [comment]
@davidberenstein1957 
I've previously attempted similar efforts to try and figure out how "large" some arbitrary Python objects are, and I came to the conclusion that `sys.getsizeof` is a really poor way to measure the size of an object, as it only works consistently with builtin objects. See this example:
```python
>>> import torch
>>> import sys
>>> vector = torch.rand(10000) 
>>> vector
tensor([0.3660, 0.8881, 0.5079,  ..., 0.2980, 0.9991, 0.7953])
>>> sys.getsizeof(vector)
72
>>> sys.getsizeof("Just an arbitrary string, for comparison purposes")
98
```

Personally, I ended up using `len(pickle.dumps(obj))`, but I used it to measure the memory usage of torch Tensors and large lists in my efforts to minimize the memory usage of various NER models. That said, it seems unnecessarily expensive to call `len(pickle.dumps(obj))` on all records.

In conclusion, I think this PR's approach won't work well, but I don't have an easy and cheap alternative for you, I'm afraid.

- Tom Aarsen

**Comment by davidberenstein1957** [comment]
That is why I was using the `json` of the records. I assumed this would correspond to the payload being sent. Anyhow, In this case, I did not care about an exact approach,  but one that would equally divide the records based on a comparable way of measuring. Not sure what the overhead is in terms of time though.

**Comment by davidberenstein1957** [comment]
@frascuchon Do you see a quick fix here? Otherwise, we can leave it for now.

**Comment by frascuchon** [comment]
> @frascuchon Do you see a quick fix here? Otherwise, we can leave it for now.

I agree with @davidberenstein1957 that measuring the json representation is a good approach for the solution. Not sure the impact of this extra computation for each record, but we can do some empirical tests.

Regarding the record size, could we use just the length of generated json string and distribute data based on that sort? 

cc @davidberenstein1957 @tomaarsen

**Comment by tomaarsen** [comment]
Considering that we convert to stdlib Python objects via the json call, the `sys.getsizeof` may still work fine, it just needs some quick checks to verify that it does indeed correctly understand that a long list corresponds with a lot of bytes. Otherwise, the string approach may work too.

File: src/argilla/client/client.py [hunk 1]
```diff
@@ -391,6 +390,18 @@ async def log_async(
...
+        # divide records in chunks of chunk size based on their byte size
+        batch_size = int(batch_size / (len(records) / batch_size)) + 1
+        n_batches = math.ceil(len(records) / batch_size)
+        rec_size = [sys.getsizeof(rec.json()) for rec in records]
+        records = [rec for _, rec in sorted(zip(rec_size, records), key=lambda pair: pair[0])]
+        records_new = dict.fromkeys(range(n_batches + 1), [])
+        while records:
+            for i in records_new:
+                records_new[i].append(records.pop(0))
+        records = [item for sublist in records_new.values() for item in sublist]
```
Comment (frascuchon) [inline src/argilla/client/client.py]: Using the modulo operator (%) we can iterate safety the records list to build the distributed batches dict.

```suggestion
        batch_size = int(batch_size / (len(records) / batch_size)) + 1
        n_batches = math.ceil(len(records) / batch_size)
        
        if n_batches > 1:
            rec_size = [sys.getsizeof(rec.json()) for rec in records]
            records = [rec for _, rec in sorted(zip(rec_size, records), key=lambda pair: pair[0])]
            
            records_new = dict.fromkeys(range(n_batches), [])
            for idx, record in enumerate(records):
                records_new[idx % n_batches].append(record)
            
            records = []
            for batch_list in records_new.values():
                records.extend(batch_list)
```

Comment (davidberenstein1957) [reply src/argilla/client/client.py]: Hi @frascuchon good suggestion. I reasoned wrongly about using the modulo operator, but gladto see it is working.
